### PR TITLE
Use docker on travis plus pipenv

### LINKED ISSUE #29: Switch to use pipenv instead of requirements.txt
As suggested in https://github.com/inveniosoftware/cookiecutter-invenio-instance/issues/4 , we could start using the [pipenv](https://docs.pipenv.org/), which seems to be becoming the standard for dependencies management in Python projects, instead of the `requirements.txt` file. It's not urgent though, as the current workflow with `requirements.txt` works fine for now.

### PR DESCRIPTION
**Pull request by egabancho**
I am still working on it, but travis seems to be happy so far. 
(Once we are happy I'll clean up the commit history and add coauthors 😉)

Should close #29 and #45 

---

Inspired by [WIP] travis: use docker-compose services #51

### CODE REVIEW

File: {{cookiecutter.project_shortname}}/docker/wait-for-services.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        if [ $RETRIES -eq 0 ]
+        then
+            echo "Couldn't reach $1"
+            exit 1
+        fi
+    done
+}
+
+{%- if cookiecutter.database == 'postgresql'%}
+_db_check(){ docker-compose exec --user postgres db bash -c "psql -c '\l'" &>/dev/null; }
```
Comment (egabancho) [inline {{cookiecutter.project_shortname}}/docker/wait-for-services.sh]: If somebody knows a better way to do this ....

File: {{cookiecutter.project_shortname}}/docker/wait-for-services.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+check_ready "MySQL" _db_check
+{%- endif %}
+
+_es_check(){ [[ $(curl -sL -w "%{http_code}\\n" "http://localhost:9200/" -o /dev/null)==200 ]]; }
+check_ready "Elasticsearch" _es_check
+
+_redis_check(){ [[ $(docker-compose exec cache bash -c "redis-cli ping")=="PONG" ]]; }
+check_ready "redis" _redis_check
+
+_rabbit_check(){ docker-compose exec mq bash -c "rabbitmqctl status" &>/dev/null; }
```
Comment (egabancho) [inline {{cookiecutter.project_shortname}}/docker/wait-for-services.sh]: If somebody knows a better way to do the check ....

File: {{cookiecutter.project_shortname}}/docker/wait-for-services.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            exit 1
+        fi
+    done
+}
+
+{%- if cookiecutter.database == 'postgresql'%}
+_db_check(){ docker-compose exec --user postgres db bash -c "psql -c '\l'" &>/dev/null; }
+check_ready "postgres" _db_check
+{%- elif cookiecutter.database == 'mysql'%}
+_db_check(){ docker-compose exec db bash -c "" &>/dev/null; }
```
Comment (egabancho) [inline {{cookiecutter.project_shortname}}/docker/wait-for-services.sh]: If somebody has any suggestion on who to do this check, otherwise I'll add `select Version();`

File: {{cookiecutter.project_shortname}}/run-tests.sh [hunk 4]
```diff
@@ -1,7 +0,0 @@
-#!/usr/bin/env bash
```
Comment (egabancho) [inline {{cookiecutter.project_shortname}}/run-tests.sh]: I (re)move this file to `scripts/test` just to follow completely https://github.com/github/scripts-to-rule-them-all, but this is just opinion and it can be change back to how it was, no hard feelings 😉

File: {{cookiecutter.project_shortname}}/scripts/server [hunk 5]
```diff
@@ -1,8 +1,13 @@
...
-    --key "$script_path"/../docker/nginx/test.key
+FLASK_ENV=development
+# Start Worker and send it to background
+pipenv run celery worker -A invenio_app.celery -l INFO & \
+    # Start server in parallel and wait for Ctrl+c
+pipenv run invenio run \
+       --cert "$script_path"/../docker/nginx/test.crt \
+       --key "$script_path"/../docker/nginx/test.key && \
+    # Once the worker is killed (Ctrl + c) kill also the server
+    kill $!
```
Comment (egabancho) [inline {{cookiecutter.project_shortname}}/scripts/server]: I like lunching Celery and the Flask server at the same time, but again, my opinion, we can revert the changes

File: {{cookiecutter.project_shortname}}/docker/wait-for-services.sh (hunk 1 above)
Comment (diegodelemos) [reply {{cookiecutter.project_shortname}}/docker/wait-for-services.sh]: There is [`pg_isready`](https://www.postgresql.org/docs/9.3/static/app-pg-isready.html), might be useful here

**Comment by egabancho** [comment]
I still need to check if the resulting application plays nicely with travis once pushed to github.

File: {{cookiecutter.project_shortname}}/.travis.yml [hunk 6]
```diff
@@ -42,41 +35,21 @@ env:
     - REQUIREMENTS=qa
 
 python:
-  - "3.6"
+  - "3.5"
```
Comment (slint) [inline {{cookiecutter.project_shortname}}/.travis.yml]: https://m.imgur.com/t/reaction/vrsLBmp

Joke aside, Python 3.5 has a similar EOL to 2.7 so maybe it's worth pushing for 3.6...

File: {{cookiecutter.project_shortname}}/scripts/server (hunk 5 above)
Comment (slint) [reply {{cookiecutter.project_shortname}}/scripts/server]: I guess the only worry is that if the app crashes ungracefully the celery worker will keep running lost in the background... I don't know if some bash (or other) trick can help with that

File: {{cookiecutter.project_shortname}}/docker/wait-for-services.sh (hunk 1 above)
Comment (egabancho) [reply {{cookiecutter.project_shortname}}/docker/wait-for-services.sh]: I know about this, but it doesn't seem to be installed in the default docker image, I'll look a bit more into it.

File: {{cookiecutter.project_shortname}}/.travis.yml (hunk 6 above)
Comment (egabancho) [reply {{cookiecutter.project_shortname}}/.travis.yml]: I know, I use this one because it's 3.5 everywher! cookiecutter `.travis`, `Dokerfile`, `setup.py` ...

**Comment by egabancho** [comment]
Sure, will do! 
Now I am trying to understand why `nvm install 6; nvm use 6` works fine if I put it inside the `.travis.yml` file, but doesn't if I put it inside `scripts/cibuild` and the call if from `.travis.yml` (`./scripts/cibuild: line 25: nvm: command not found`), any ideas on that?
